Lib/multiprocessing/synchronize.py

```diff
@@ -328,41 +328,70 @@ def wait_for(self, predicate, timeout=None):
+        # If the thread interrupted by the signal handler is wait()-ing and the signal handler calls set(),
+        # this is fine as wait() spins on the value. Thus, after the signal handler is done, the thread will
+        # return from wait()
+        # Fixes https://github.com/python/cpython/issues/85772
+        with self._flag:
+            with self._set_id:
+                if self._flag.value == 0:
+                    # There is a theoretical chance of race here. It requires the following conditions:
+                    # The interrupted thread must be wait()ing.
+                    # Then set must be called reentrant for the maximum value of c_ulonglong times,
+                    # and all interruptions must happen exactly after `if self._flag.value == 0:`.
+                    # The _set_id value will then wrap around. Then clear() must be called
+                    # before the original wait() code continues. The wait() code will then continue
+                    # to (incorrectly) wait. I think this case is safe to ignore. The stack
+                    # will grow too large before there is any chance of this actually happening.
+
+                    self._flag.value = 1
+                    self._set_id.value += 1
+                    # There is no race here by reentrant set when reaching the maximum value for `self._set_id.value`.
+                    # ctypes.c_ulonglong will overflow without any exception:
+                    # https://docs.python.org/3/library/ctypes.html#ctypes.c_ulonglong
+                    # > no overflow checking is done.
+                    # This means that we do not need to check if some maximum value is reached:
+                    # C will wrap around the value for us.
 
     def clear(self):
-        with self._cond:
-            self._flag.acquire(False)
+        with self._flag:
+            self._flag.value = 0
 
     def wait(self, timeout=None):
-        with self._cond:
-            if self._flag.acquire(False):
-                self._flag.release()
-            else:
-                self._cond.wait(timeout)
-
-            if self._flag.acquire(False):
-                self._flag.release()
+        start_time = time.monotonic()
+        set_id = self._set_id.value
+        while True:
+            if self._flag.value == 1:
                 return True
-            return False
+            elif set_id != self._set_id.value:
+                return True # flag is unset, but set_id changed, so there must have been a `set` followed by a `clear`
+                            # during `time.sleep()`. Fixes https://github.com/python/cpython/issues/95826
+            elif timeout is not None and (time.monotonic() - start_time) > timeout:
+                return False
+            else:
+                # Fixes https://github.com/python/cpython/issues/85772 by spinning and sleeping.
```

Comment:
Iff (lets _not_ assume we should go this route - see my comment on the issue) we're going to abandon use of the native OS platform APIs which properly implement timeouts on inter-process semaphores without a busy loop via `_multiprocessing.SemLock` in `Modules/_multiprocessing/semaphore.c`, sleeping in the loop should be done in an exponential back-off fashion as was the case even in `threading` itself before we started using proper OS APIs there.  See https://github.com/python/cpython/blob/v3.1.3/Lib/threading.py#L227 for the old exponential back-off delay example.

Doing a busy loop with sleeps as a low level primitive in 2024 feels **very** wrong to me. They've always been _really unfriendly_ to both latency due to unnecessary delays and power usage from frequent unnecessary wakes.

I suggested an alternate idea in #126434.

#### History

Prior to CPython 3.2 `threading.Condition` was implemented with a back-off in a similar manner.  3.2 improved on that old hack by using the OS APIs for lock timeouts in https://github.com/python/cpython/commit/7c3e5773954009d65520eb063621cf7724da88e3

Reply:
Thanks @gpshead and @ZeroIntensity 

I will read your comments and suggestions thoroughly tomorrow, and come back to you.

Thanks and kind regards.